flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/aggregate/PercentileAggFunction.java

```diff
@@ -0,0 +1,382 @@
+
+    protected final transient DataType valueType;
+    protected final transient DataType frequencyType;
+
+    public PercentileAggFunction(LogicalType inputType, LogicalType frequencyType) {
+        this.valueType = toInternalDataType(inputType);
+        this.frequencyType = frequencyType == null ? null : toInternalDataType(frequencyType);
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Accumulator
+    // --------------------------------------------------------------------------------------------
+
+    /** Accumulator for PERCENTILE. */
+    public static class PercentileAccumulator {
+
+        public double[] percentages;
+        public long totalCount;
+        public MapView<Double, Long> valueCount;
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+            PercentileAccumulator that = (PercentileAccumulator) o;
+            return Arrays.equals(percentages, that.percentages)
+                    && totalCount == that.totalCount
+                    && Objects.equals(valueCount, that.valueCount);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(Arrays.hashCode(percentages), totalCount, valueCount.hashCode());
+        }
+
+        public Double[] getValue() {
+            List<Pair<Double, Integer>> sortedPercentages =
+                    IntStream.range(0, percentages.length)
+                            .mapToObj(
+                                    index ->
+                                            new Pair<>(
+                                                    percentages[index] * (totalCount + 1), index))
+                            .sorted(Comparator.comparing(Pair::getKey))
+                            .collect(Collectors.toList());
+
+            List<Map.Entry<Double, Long>> sortedList = new ArrayList<>();
+            try {
+                for (Map.Entry<Double, Long> entry : valueCount.entries()) {
+                    sortedList.add(entry);
+                }
+            } catch (Throwable t) {
+                t.printStackTrace();
+            }
+
+            sortedList =
+                    sortedList.stream()
+                            .sorted(Map.Entry.comparingByKey())
+                            .collect(Collectors.toList());
+
+            Double[] percentiles = new Double[percentages.length];
+
+            long preCnt = sortedList.get(0).getValue();
+            for (int i = 0, j = 0; i < sortedPercentages.size(); i++) {
+                double position = sortedPercentages.get(i).getKey();
+                long lower = (long) Math.floor(position);
+                long higher = (long) Math.ceil(position);
+
+                while (preCnt < lower) {
+                    j++;
```

Comment:
Should we prevent `j` out of bound of the `sortedList`?

Reply:
You are right, I forgot that `lower` can be `totalCount + 1` when `percentage = 1.0`.